Source/Provers/SMTLib/SMTLibProcess.cs

```diff
@@ -135,11 +132,78 @@ public override void Send(string cmd)
       }
 
       toProver.WriteLine(cmd);
+      toProver.Flush();
     }
 
-    internal Inspector Inspector => inspector;
+    // TODO, consider building in a PingPong to catch a missing response.
+    public override async Task<SExpr> SendRequest(string request) {
+      SExpr previousResponse = null;
+      try {
+        await previousRequestsAreDone.WaitAsync();
+        Send(request);
+        Send(PingRequest);
+        while (true) {
+          var response = await GetProverResponse();
+          if (IsPong(response)) {
+            if (previousResponse == null) {
+              throw new Exception("Request returned no response");
+            }
+            return previousResponse;
+          }
+          previousResponse = response;
+        }
+      }
+      finally {
+        previousRequestsAreDone.Release();
+      }
+    }
+
+    public override async Task PingPong() {
+      try {
+        await previousRequestsAreDone.WaitAsync();
+        Send(PingRequest);
+        while (true) {
+          var response = await GetProverResponse();
+          if (response == null)
+          {
+            throw new ProverDiedException();
+          }
+
+          if (IsPong(response))
+          {
+            return;
+          }
+
+          HandleError("Invalid PING response from the prover: " + response.ToString());
+        }
+      }
+      finally {
+        previousRequestsAreDone.Release();
+      }
+    }
+
+    public override async Task<IReadOnlyList<SExpr>> SendRequestsAndCloseInput(IReadOnlyList<string> requests) {
```

Comment:
The name here suggests that it'll close the input handle of the solver before returning, but I don't see how it does that.

Reply:
Woops, I accidentally removed that, but brought it back now.